### PR TITLE
[casr-python][casr-java] Add sub-tool option

### PR DESCRIPTION
**Pull request by hkctkuy**

### CODE REVIEW

File: casr/src/bin/casr-java.rs [hunk 1]
```diff
@@ -56,6 +56,17 @@ fn main() -> Result<()> {
                 .value_name("FILE")
                 .help("File with regular expressions for functions and file paths that should be ignored"),
         )
+        .arg(
+            Arg::new("sub-tool")
+                .long("sub-tool")
+                .default_value("casr-san")
+                .action(ArgAction::Set)
+                .value_parser(clap::value_parser!(PathBuf))
```
Comment (SweetVishnya) [inline casr/src/bin/casr-java.rs:64]: validate option, search in PATH

File: casr/src/bin/casr-java.rs [hunk 2]
```diff
@@ -56,6 +56,17 @@ fn main() -> Result<()> {
...
+                .value_name("BIN")
+                .help(
+                    "Path to sub tool bin for crash analyze that will be called in case main tool fails",
```
Comment (SweetVishnya) [inline casr/src/bin/casr-java.rs:67]: remove `bin`

Comment (SweetVishnya) [reply casr/src/bin/casr-java.rs:67]: analyze -> analysis

Comment (SweetVishnya) [reply casr/src/bin/casr-java.rs:67]: called when main tool fails to detect a crash

File: casr/src/bin/casr-python.rs [hunk 3]
```diff
@@ -55,6 +55,17 @@ fn main() -> Result<()> {
...
         )
+        .arg(
+            Arg::new("sub-tool")
+                .long("sub-tool")
+                .default_value("casr-san")
+                .action(ArgAction::Set)
+                .value_parser(clap::value_parser!(PathBuf))
+                .value_name("BIN")
+                .help(
+                    "Path to sub tool bin for crash analyze that will be called in case main tool fails",
```
Comment (SweetVishnya) [inline casr/src/bin/casr-python.rs:66]: The same comments as in casr-java

File: casr/src/bin/casr-java.rs (hunk 1 above)
Comment (SweetVishnya) [reply casr/src/bin/casr-java.rs:64]: You can also parse the absolute path here

File: casr/src/util.rs [hunk 4]
```diff
@@ -17,17 +17,18 @@ use std::path::{Path, PathBuf};
...
 ///
 /// * `matches` - casr options
 ///
-/// * `tool` - tool, that called casr-san
+/// * `tool` - tool, that called sub tool
 ///
 /// * `argv` - executable file options
-pub fn call_casr_san(matches: &ArgMatches, argv: &[&str], tool: &str) -> Result<()> {
-    let mut cmd = Command::new("casr-san");
+pub fn call_sub_tool(matches: &ArgMatches, argv: &[&str], tool: &str) -> Result<()> {
```
Comment (SweetVishnya) [inline casr/src/util.rs:29]: we'd better pass sub tool in `tool` argument

File: casr/src/util.rs [hunk 5]
```diff
@@ -50,7 +51,7 @@ pub fn call_casr_san(matches: &ArgMatches, argv: &[&str], tool: &str) -> Result<
     if output.status.success() {
         Ok(())
     } else {
-        bail!("casr-san error when calling from {tool}");
+        bail!("{sub_tool:?} error when calling from {tool}");
```
Comment (SweetVishnya) [inline casr/src/util.rs:54]: sub_tool -> tool

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: casr/src/bin/casr-java.rs (hunk 2 above)
Comment (hkctkuy) [reply casr/src/bin/casr-java.rs]: Done

File: casr/src/bin/casr-python.rs (hunk 3 above)
Comment (hkctkuy) [reply casr/src/bin/casr-python.rs]: Done

File: casr/src/bin/casr-java.rs (hunk 1 above)
Comment (hkctkuy) [reply casr/src/bin/casr-java.rs:64]: Done

File: casr/src/util.rs (hunk 4 above)
Comment (hkctkuy) [reply casr/src/util.rs]: Done

File: casr/src/util.rs (hunk 5 above)
Comment (hkctkuy) [reply casr/src/util.rs]: Done

File: casr/src/bin/casr-java.rs [hunk 6]
```diff
@@ -56,6 +56,17 @@ fn main() -> Result<()> {
                 .value_name("FILE")
                 .help("File with regular expressions for functions and file paths that should be ignored"),
         )
+        .arg(
+            Arg::new("sub-tool")
+                .long("sub-tool")
+                .default_value("casr-san")
+                .action(ArgAction::Set)
+                .value_parser(clap::value_parser!(PathBuf))
+                .value_name("BIN")
```
Comment (SweetVishnya) [inline casr/src/bin/casr-java.rs:65]: BIN -> PATH

File: casr/src/bin/casr-python.rs [hunk 7]
```diff
@@ -55,6 +55,17 @@ fn main() -> Result<()> {
...
         )
+        .arg(
+            Arg::new("sub-tool")
+                .long("sub-tool")
+                .default_value("casr-san")
+                .action(ArgAction::Set)
+                .value_parser(clap::value_parser!(PathBuf))
+                .value_name("BIN")
+                .help(
+                    "Path to sub tool for crash analysis that will be called when main tool fails to detect a crash",
```
Comment (SweetVishnya) [inline casr/src/bin/casr-python.rs:66]: Update usage docs

File: casr/src/util.rs [hunk 8]
```diff
@@ -13,21 +13,35 @@ use simplelog::*;
...
-pub fn call_casr_san(matches: &ArgMatches, argv: &[&str], tool: &str) -> Result<()> {
-    let mut cmd = Command::new("casr-san");
+pub fn call_sub_tool(matches: &ArgMatches, argv: &[&str], name: &str) -> Result<()> {
+    let tool = matches.get_one::<PathBuf>("sub-tool").unwrap();
+    if let Err(_) = which(tool) {
+        if !tool.exists() {
+            bail!("Sub tool {tool:?} doesn't exist");
+        }
+        if !tool.is_file() {
+            bail!("Sub tool {tool:?} isn't exist");
```
Comment (SweetVishnya) [inline casr/src/util.rs:38]: doesn't

Comment (hkctkuy) [reply casr/src/util.rs:38]: Done

File: casr/src/bin/casr-java.rs (hunk 6 above)
Comment (hkctkuy) [reply casr/src/bin/casr-java.rs:65]: Done

File: casr/src/bin/casr-python.rs (hunk 7 above)
Comment (hkctkuy) [reply casr/src/bin/casr-python.rs:66]: Done

File: casr/src/util.rs [hunk 9]
```diff
@@ -35,7 +35,7 @@ pub fn call_sub_tool(matches: &ArgMatches, argv: &[&str], name: &str) -> Result<
             bail!("Sub tool {tool:?} doesn't exist");
         }
         if !tool.is_file() {
-            bail!("Sub tool {tool:?} isn't exist");
+            bail!("Sub tool {tool:?} isn't file");
```
Comment (SweetVishnya) [inline casr/src/util.rs:38]: a file

File: docs/usage.md [hunk 10]
```diff
@@ -117,6 +117,8 @@ Create CASR reports (.casrep) from python reports
           --stdin <FILE>     Stdin file for program
           --ignore <FILE>    File with regular expressions for functions and file paths that
                              should be ignored
+          --sub-tool <BIN>   Path to sub tool bin for crash analyze that will be called in
```
Comment (SweetVishnya) [inline docs/usage.md:120]: PATH

Comment (SweetVishnya) [reply docs/usage.md:120]: old usage, rebuil

File: casr/src/util.rs (hunk 9 above)
Comment (hkctkuy) [reply casr/src/util.rs]: Done

**Review by Avgor46 — APPROVED** [review]

**Review by SweetVishnya — APPROVED** [review]

**Review by anfedotoff — APPROVED** [review]

**Pull request merged**